third_party/move/move-compiler-v2/src/acquires_checker.rs

```diff
@@ -0,0 +1,258 @@
+                .collect();
+            acquire_env.insert(fun_id, AcquiredResources(acquired));
+        }
+        (call_graph, acquire_env)
+    }
+
+    fn compute_fixed_point(
+        &self,
+        call_graph: BTreeMap<FunId, BTreeMap<FunId, Loc>>,
+        mut acquire_env: BTreeMap<FunId, AcquiredResources>,
+    ) -> BTreeMap<FunId, AcquiredResources> {
+        let reversed_call_graph = call_graph.iter().fold(
+            BTreeMap::new(),
+            |mut reversed: BTreeMap<FunId, BTreeSet<FunId>>, (caller, callees)| {
+                for (callee, _) in callees.iter() {
+                    reversed.entry(*callee).or_default().insert(*caller);
+                }
+                reversed
+            },
+        );
+        let mut work_list: VecDeque<_> = self
+            .mod_env
+            .get_functions()
+            .map(|fun_env| fun_env.get_id())
+            .collect();
+        while let Some(fun_id) = work_list.pop_front() {
+            let mut any_changes = false;
+            let mut caller_acquires = acquire_env.remove(&fun_id).expect("acquired resources");
+            for (callee, loc) in call_graph.get(&fun_id).expect("callees") {
+                if *callee == fun_id {
+                    continue;
+                }
+                let callee_acquires = acquire_env.get(callee).expect("callee acquires");
+                let changed = caller_acquires.join(*callee, loc.clone(), callee_acquires);
+                any_changes = any_changes || changed;
+            }
+            acquire_env.insert(fun_id, caller_acquires);
+            if any_changes {
+                if let Some(callers) = reversed_call_graph.get(&fun_id) {
+                    for caller in callers {
+                        work_list.push_back(*caller);
+                    }
+                }
+            }
+        }
+        acquire_env
+    }
+}
+
+/// Suppose the given function is defined in module M.
+/// Returns
+/// - the calles of the given function that are defined in M
+/// - resources acquired by move_from\<T>, borrow_global_mut\<T>, or borrow_global\<T>,
+/// where T is define in M
+fn get_callees_and_acquired_resources(
+    fun_env: FunctionEnv,
+) -> (BTreeMap<FunId, Loc>, BTreeMap<StructId, Loc>) {
+    let mut callees = BTreeMap::new();
+    let mut resources = BTreeMap::new();
+    let mid = fun_env.module_env.get_id();
+    if let Some(fun_body) = fun_env.get_def() {
+        let mut collect_callees = |exp: &ExpData| match exp {
+            ExpData::Call(node_id, op, _) => {
+                if let Operation::MoveFunction(exp_mid, exp_fid) = op {
+                    if *exp_mid == fun_env.module_env.get_id() {
+                        let loc = fun_env.module_env.env.get_node_loc(*node_id);
+                        callees.entry(*exp_fid).or_insert(loc);
+                    }
+                }
+                true
+            },
+            ExpData::SpecBlock(..) => false,
```

Comment:
Sorry guys, I found this bug and spent some time fixing it. I was trying to skip visiting spec blocks and all their subexpression, but this short-circuits the entire visit. This has been fixed in the latest commit.

Reply:
For ease of comparison,
before
```
        let mut collect_callees = |exp: &ExpData| match exp {
            ExpData::Call(node_id, op, _) => {
                if let Operation::MoveFunction(exp_mid, exp_fid) = op {
                    if *exp_mid == fun_env.module_env.get_id() {
                        let loc = fun_env.module_env.env.get_node_loc(*node_id);
                        callees.entry(*exp_fid).or_insert(loc);
                    }
                }
                true
            },
            ExpData::SpecBlock(..) => false, // WRONG!
            _ => true,
        };
```
after
```
        let mut in_spec_block = false;
        let mut collect_callees = |pos, exp: &ExpData| match exp {
            ExpData::Call(node_id, op, _) => {
                if !in_spec_block && matches!(pos, VisitorPosition::Pre) {
                    if let Operation::MoveFunction(exp_mid, exp_fid) = op {
                        if *exp_mid == fun_env.module_env.get_id() {
                            let loc = fun_env.module_env.env.get_node_loc(*node_id);
                            callees.entry(*exp_fid).or_insert(loc);
                        }
                    }
                }
                true
            },
            ExpData::SpecBlock(..) => {
                match pos {
                    VisitorPosition::Pre => in_spec_block = true,
                    VisitorPosition::Post => in_spec_block = false,
                    _ => {},
                }
                true
            },
            _ => true,
        };
```